### PR TITLE
Put program list in the cache and validate the program id in the URL.

### PR DESCRIPTION
**Pull request by toddburnside**

### CODE REVIEW

File: common/src/main/scala/explore/cache/ProgramCache.scala [hunk 1]
```diff
@@ -205,7 +217,17 @@ object ProgramCache extends CacheComponent[ProgramSummaries, ProgramCache]:
...
+            ProgramSummaries.programs
+              .modify(_.updated(data.programEdit.value.id, data.programEdit.value))
+          )
+        )
+
     // TODO Handle errors, disable transparent resubscription upon connection loss.
-    (updateTargets, updateObservations, updateGroups, updateAttachments).mapN(
-      _.merge(_).merge(_).merge(_)
+    (updateTargets, updateObservations, updateGroups, updateAttachments, updatePrograms).mapN(
+      _.merge(_).merge(_).merge(_).merge(_)
```
Comment (cquiroz) [inline common/src/main/scala/explore/cache/ProgramCache.scala:232]: lots of merging

File: explore/src/main/scala/explore/Routing.scala [hunk 2]
```diff
@@ -102,20 +116,25 @@ object Routing:
...
+      SequenceEditor(routingInfo.programId)
+    )
 
   private def showProgramSelectionPopup(model: View[RootModel]): VdomElement =
-    ProgramsPopup(
-      currentProgramId = none,
-      undoStacks = model.zoom(RootModel.undoStacks)
-    )
+    // Because we are not supplying a program id, the ProgramsPopup will be displayed
+    withProgramSummaries(none, model)(_ => <.div("Programmer error!"))
```
Comment (cquiroz) [inline explore/src/main/scala/explore/Routing.scala:137]: Cool, I'm looking forward to the first faul report with "Programmer error"

**Comment by cquiroz** [comment]
So, if the program id doesn't exist (or not allowed) the program popup gets shown?

File: common/src/main/scala/explore/cache/ProgramCache.scala (hunk 1 above)
Comment (toddburnside) [reply common/src/main/scala/explore/cache/ProgramCache.scala:232]: Is there a better way?

File: explore/src/main/scala/explore/Routing.scala (hunk 2 above)
Comment (toddburnside) [reply explore/src/main/scala/explore/Routing.scala:137]: Maybe I should add my name? 😆

File: common/src/main/scala/explore/cache/ProgramCache.scala (hunk 1 above)
Comment (cquiroz) [reply common/src/main/scala/explore/cache/ProgramCache.scala:232]: not sure what merge does, maybe `foldLeft`?

**Comment by toddburnside** [comment]
> So, if the program id doesn't exist (or not allowed) the program popup gets shown?

Yep. And if there was an id in the url, a message is displayed at the bottom of the popup. 
image
 Because of the way the program cache works, if the program is deleted the user will be immediately taken to the popup. As long as the programEdit subscription works for it, the same thing will happen if authorization is revoked.

**Review by cquiroz — APPROVED** [review]
LGTM, this is good

**Comment by toddburnside** [comment]
> As long as the programEdit subscription works for it, the same thing will happen if authorization is revoked.

Actually, we'll probably need to do a bit more to get authorization revoking to work right. We'll need some more information in the subscription, I think.

File: common/src/main/scala/explore/cache/ProgramCache.scala (hunk 1 above)
Comment (toddburnside) [reply common/src/main/scala/explore/cache/ProgramCache.scala:232]: > not sure what merge does, maybe `foldLeft`?

Raul thumbed up this comment. I'm not sure what that means.

**Review by rpiaggio — APPROVED** [review]
LGTM

File: common/src/main/scala/explore/cache/ProgramCache.scala (hunk 1 above)
Comment (rpiaggio) [reply common/src/main/scala/explore/cache/ProgramCache.scala:232]: That we can use `List(...).reduceLeft(_ merge _)`.

Comment (toddburnside) [reply common/src/main/scala/explore/cache/ProgramCache.scala:232]: Ah. I change that.

Comment (toddburnside) [reply common/src/main/scala/explore/cache/ProgramCache.scala:232]: Actually, I think it was simpler with `sequence.map` in this case.

**Review by rpiaggio — APPROVED** [review]
LGTM

**Pull request merged**